### PR TITLE
Consul eds

### PR DESCRIPTION
**Pull request by yuval-k**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/2567

### CODE REVIEW

File: projects/gloo/pkg/plugins/consul/eds.go [hunk 1]
```diff
@@ -166,6 +172,7 @@ func buildEndpointsFromSpecs(ctx context.Context, writeNamespace string, resolve
 	var endpoints v1.EndpointList
 	for _, spec := range specs {
 		if upstreams, ok := trackedServiceToUpstreams[spec.ServiceName]; ok {
+			// TODO if buildEndpoints fails temporarily due to dns failure, we will remove it from eds.
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/consul/eds.go]: let's open an issue to track this instead

File: projects/gloo/pkg/plugins/consul/eds_test.go [hunk 2]
```diff
@@ -450,6 +459,151 @@ var _ = Describe("Consul EDS", func() {
...
+				{
+					ID:         "12341234-1234-1234-1234-123412341234",
+					Node:       "ip-1.2.3.4",
+					Address:    "1.2.3.4",
+					Datacenter: "test",
+					TaggedAddresses: map[string]string{
+						"lan": "1.2.3.4",
+						"wan": "1.2.3.4",
+					},
+					// test with two services having the same services id. this can happen.
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/consul/eds_test.go]: per https://www.consul.io/docs/agent/services.html

> It is required that all services have a unique ID per node, so if names might conflict then unique IDs should be provided.

If this is not true, as you saw in testing (same service ID but different ports), then we need to update the consul spec collector as well: https://github.com/solo-io/gloo/blob/v1.3.12/projects/gloo/pkg/plugins/consul/eds.go#L368-L394

File: changelog/v1.3.13/consul-eds.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+changelog:
+  - type: FIX
+    description: Fix issue where Consul endpoints were given a similar name, due to them having the same ID.
+    issueLink: https://github.com/solo-io/gloo/issues/2567
```
Comment (kdorosh) [inline changelog/v1.3.13/consul-eds.yaml]: let's add a changelog entry for this part of the PR:

> When not empty, only service instances that match all of the tags will be selected for this upstream.

File: projects/gloo/pkg/plugins/consul/eds.go (hunk 1 above)
Comment (yuval-k) [reply projects/gloo/pkg/plugins/consul/eds.go]: https://github.com/solo-io/gloo/issues/2576

File: changelog/v1.3.13/consul-eds.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  - type: FIX
+    description: When not empty, only service instances that match all of the tags will be selected for this upstream.
```
Comment (kdorosh) [inline changelog/v1.3.13/consul-eds.yaml]:
```suggestion
    description: When not empty, only consul service instances that match all of the tags will be selected for this upstream.
```

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,287 @@
+{
```
Comment (kdorosh) [inline .vscode/launch.json]: let's remove this file and add `.vscode` dir to gitignore

File: projects/gloo/api/v1/options/consul/consul.proto [hunk 6]
```diff
@@ -18,12 +18,20 @@ import "gloo/projects/gloo/api/v1/options/service_spec.proto";
 message UpstreamSpec {
     // The name of the Consul Service
     string service_name = 1;
+
+    // Deprecated: the meaning of this field is ambigous. please use subset_tags and instance_tags instead.
+    repeated string service_tags = 2;
```
Comment (kdorosh) [inline projects/gloo/api/v1/options/consul/consul.proto]: we should define what this field actually does

File: .vscode/launch.json (hunk 5 above)
Comment (yuval-k) [reply .vscode/launch.json]: right, missed that - thanks!

**Pull request merged**